feat: design submission form

Midnighter opened the pull request:
WIP:

* Still need to add in new project addition option
* Need to select a model based on organism
* Make actual request to metabolic-ninja

Question:

* I have a commented line because the `AxiosResponse` type annotation doesn't work. Does anyone know why?

Midnighter commented on src/components/Design.vue:
```diff
@@ -0,0 +1,183 @@
+    return {
+      isValid: false,
+      organism: "",
+      organismRules: [v => !!v || "Selecting an organism is required."],
```
Maybe... Although the error messages are special and in this case I'm more on the side of @kvikshaug that as long as the message is different I prefer duplication. Could use a more general "Field is required", though.

ChristianLieven replied:
If you fo with the more general "Field is required" or simply "Required" then you might as well define it only once.

phantomas1234 commented:
@Midnighter can you please avoid ec7 becoming the default yeast model if you can because it doesn't work. We should integrate the yeast 8 model that has bigg IDs soon.

Midnighter commented on src/components/Models.vue:
```diff
@@ -264,7 +264,7 @@ export default Vue.extend({
       rowsPerPage: 10
     },
-    show: false,
+    showAdvanced: false,
```
Yes, edited wrong file.

Midnighter commented:
I have now completed the design submission form. The mapping started in `src/store/modules/models.ts` needs to be extended. Also the model selection method could partially be moved to the models store, I guess.

I'd like to merge this for now since I need to work on the pitch.

ChristianLieven commented on src/components/Design.vue:
```diff
@@ -0,0 +1,341 @@
+          this.$store.commit("setFetchError", error, { root: true });
+        });
+    },
+    selectModels(project: ProjectItem, organism: OrganismItem): void {
```
This didn't seem to work for me?

![Screenshot 2019-05-10 at 13 26 04](https://user-images.githubusercontent.com/8027271/57524130-2ef15f00-7327-11e9-875f-90210565220f.png)

Midnighter commented on src/components/Design.vue:
```diff
@@ -0,0 +1,341 @@
+<template xmlns:v-slot="http://www.w3.org/1999/XSL/Transform">
```
This comes from Webstorm being confused by Vue syntax. It thinks `v-slot:xmlns` is an XML declaration and thus added this import :/ Thanks for catching that.

Midnighter commented on src/components/Design.vue:
```diff
@@ -0,0 +1,341 @@
+  beforeRouteLeave(to: Route, from: Route, next: NextHandler) {
+    if (!this.isSubmitted) {
+      const answer = window.confirm(
+        "Do you really want to leave? You have pending changes."
```
I hadn't noticed that. Yes, the dialog should not appear when the form is empty. IMO it's good practice to do it, though.

Midnighter commented on src/components/Design.vue:
```diff
@@ -0,0 +1,341 @@
+            <template v-slot:item="data">
+              <v-list-tile-content
+                v-text="data.item.name"
+                class="font-italic"
```
Yes, organism names should generally be italic.

Midnighter commented on src/components/Design.vue:
```diff
@@ -0,0 +1,341 @@
+            :rules="projectRules"
+            clearable
+          >
+            <template v-slot:prepend-item>
```
I'm happy to make a PR in order to modify other lists, too.

Midnighter commented on src/components/Design.vue:
```diff
@@ -0,0 +1,341 @@
+  name: string;
+}
+
+type Nullable<T> = T | null | undefined;
```
Sure, the only reason I put it here is that I wasn't aware of a more general place to put type definitions and I didn't want to start one without discussing it with everyone.

Midnighter replied:
Moved to `src/types/general.ts`.

Midnighter commented on src/components/Design.vue:
```diff
@@ -0,0 +1,341 @@
+              <v-spacer></v-spacer>
+              <v-btn icon>
+                <v-icon>{{
+                  showAdvanced ? "keyboard_arrow_down" : "keyboard_arrow_up"
```
This was copy & pasted from Vuetify. Happy to change :slightly_smiling_face:

Midnighter commented on src/components/Design.vue:
```diff
@@ -0,0 +1,341 @@
+            :rules="organismRules"
+            clearable
+            placeholder="e.g., Escherichia coli"
+            class="font-italic"
```
Cheers for catching that.

Midnighter commented on src/components/Design.vue:
```diff
@@ -0,0 +1,341 @@
+      );
+      this.model = organism2ModelMapping[this.organism.id];
+    },
+    modelRules(): RuleHandler[] {
```
I had problems with the rules not updating when `this.project` or `this.organism` change thus I made it a method for now.

Midnighter replied:
I took a different approach to the rules. It's better now.

Midnighter commented on src/components/Design.vue:
```diff
@@ -0,0 +1,341 @@
+
+type Nullable<T> = T | null | undefined;
+type RuleOutcome = string | true;
+type RuleHandler = (any) => RuleOutcome;
```
Okay, I haven't explored the type definitions provided by Vuetify. I'll take a look. If they don't provide them, though, I think we should add them but in a more general place.

Midnighter replied:
Moved to `src/types/general.ts`.

Midnighter commented on src/components/Design.vue:
```diff
@@ -0,0 +1,341 @@
+type Nullable<T> = T | null | undefined;
+type RuleOutcome = string | true;
+type RuleHandler = (any) => RuleOutcome;
+type NextHandler = (to?: RawLocation | false | Function | void) => void;
```
There will be a proper type definition in future (https://github.com/vuejs/vue-router/issues/2758), however, so far there isn't one.

Midnighter replied:
Moved to `src/types/general.ts`.

Midnighter commented on src/components/Design.vue:
```diff
@@ -0,0 +1,341 @@
+    NewModel
+  },
+  data() {
+    return {
```
Okay, that was supposed to be the point :frowning_face: maybe that means I need to define an interface and let `data()` return that?

Midnighter replied:
Alright, I have tried again with defining an interface. On one hand it's still possible to assign any other types which probably has to do with the component state in practice being something else than the object returned by `data()`. On the other hand, having the interface definition still makes auto-completion on properties possible so that's quite nice. I'd call it a nice-to-have. Additionally, defining an interface for `data()` also documents the intention for the various data types.

Midnighter commented on src/components/Design.vue:
```diff
@@ -0,0 +1,341 @@
+  created() {
+    this.fetchProducts();
+  },
+  beforeRouteLeave(to: Route, from: Route, next: NextHandler) {
```
For the `authGuard` I was able to annotate that it is of type `NavigationGuard`, i.e., a specific function type. This then already allows typescript to infer the correct types for the function arguments. However, I don't know how to do that in Vue's object notation thus the only way I saw was to annotate the arguments here.

ChristianLieven commented on src/router.ts:
```diff
@@ -1,18 +1,23 @@
 import TermsOfService from "./views/TermsOfService.vue";
+import NotFound from "./views/NotFound.vue";
+
+// FIXME (Moritz Beber): In future we would like to import this from vue-router.
```
I remember that we discussed this previously on memote source code. Wouldn't it be better to just open an issue which references this PR (or line of code if thats possible), rather than leaving a comment in the source code?

Midnighter replied:
https://github.com/DD-DeCaF/caffeine-vue/issues/40

ChristianLieven commented on src/store/modules/models.ts:
```diff
@@ -13,6 +13,10 @@ export interface ModelItem {
   model_serialized?: object;
 }
 
+export const organism2ModelMapping = {
```
Is this the yeast work-around?

Maybe you should open an issue for this too as it seems somewhat hacky and we might forget about it?

Midnighter replied:
#41

ChristianLieven commented on src/views/Design.vue:
```diff
@@ -0,0 +1,362 @@
+    projectOptions(): ProjectItem[] {
+      return this.$store.state.projects.projects;
+    },
+    // TODO (Moritz Beber): Selection of organism in future should depend on the
```
Opening an issue here will make corrections like this more manageable than TODO comments in the source code.

Midnighter replied:
#42

kvikshaug commented on src/views/Design.vue:
```diff
@@ -0,0 +1,363 @@
+      }
+    },
+    modelRules(): RuleHandler[] {
+      return [this.modelOkayRule, v => !!v || "Model is required"];
```
This doesn't quite work as expected. If you select a project, then an organism with no mapping (e.g. cricetulus griseus), then click the model select, then deselect it by clicking somewhere else, you still get the "Please first select a project and organism" error (expected "Model is required").

Midnighter replied:
My assumption was that there should always be at least one model for an organism. However, I now see that this may not always be the case.

Midnighter replied:
I'm not sure it can be done with Vuetify only. Would have to manually call validation upon the field losing focus.

kvikshaug replied:
How about having the validation only be concerned with whether the model is set, and instead use a `:hint` if project and/or organism is missing?

Midnighter commented on src/views/Design.vue:
```diff
@@ -0,0 +1,363 @@
+  watch: {
+    modelOkay(isOkay: boolean): void {
+      if (isOkay) {
+        this.model = organism2ModelMapping[this.organism.id];
```
Absolutely!

Midnighter commented:
@kvikshaug I've addressed your comments. I'm not 100% happy with the field validation yet but the hint is a decent solution for now. Not willing to spend any more time on it.

I've also removed `next` type annotation. I agree that we're not in full control of what that type actually is, so we shouldn't declare it ourselves. I've left in the types regarding rules, though.
